### PR TITLE
Allow SearchField objects to set Solr field type

### PR DESCRIPTION
**Pull request by ahoho**
Currently there's no easy way for a given `SearchField` object to explicitly set its Field Type in the Solr Schema. Obviously, this can be changed from within the Solr Admin page, but I feel that an integrated solution is preferable.

This is particularly useful for indexed fields that need not be tokenized and should support sorting (e.g., ID numbers, emails, etc.).

### CODE REVIEW

**Comment by acdha** [comment]
We should probably add some tests for this. 

I'm also wonder whether the attribute should either be clearly Solr-specific, be implemented for at least ElasticSearch, or perhaps just issue a warning / NotImplementedError when encountered by the non-Solr backends.

**Comment by ahoho** [comment]
Those are a good points. I'm not sure how to handle tests, but I'll make the changes to either specify Solr-specific compatibility or add support for Elasticsearch.

On Jan 29, 2016, at 3:50 PM, Chris Adams <notifications@github.com<mailto:notifications@github.com>> wrote:


We should probably add some tests for this.

I'm also wonder whether the attribute should either be clearly Solr-specific, be implemented for at least ElasticSearch, or perhaps just issue a warning / NotImplementedError when encountered by the non-Solr backends.

—
Reply to this email directly or view it on GitHub<https://github.com/django-haystack/django-haystack/pull/1304#issuecomment-176964075>.
******************
 
IMPORTANT NOTICE FROM THE BRATTLE GROUP: This message, and any attachments, are
intended only for the use of the individual or entity to which it is addressed and
may contain information that is privileged, confidential, and exempt from disclosure
under applicable law. Any unauthorized dissemination, distribution or copying of this
communication is strictly prohibited. If you have received this communication in error
please notify me immediately by return email and immediately delete the original and all
copies of the message and any attachments to it. Note also that nothing in this message is
intended to constitute an electronic signature or otherwise to satisfy the requirements for
a contract unless an express statement to the contrary is included in the message.

Please ensure you have adequate virus protection before you open or use attachments.
The Brattle Group does not accept any liability for viruses.
 
******************

**Comment by ahoho** [comment]
@acdha, this patch now raises a `NotImplementedError` when using a non-Solr backend. I'm still now sure how best to test (outside of confirming that it works on my own project).

File: haystack/fields.py [hunk 1]
```diff
@@ -26,7 +26,7 @@ class SearchField(object):
     def __init__(self, model_attr=None, use_template=False, template_name=None,
                  document=False, indexed=True, stored=True, faceted=False,
                  default=NOT_PROVIDED, null=False, index_fieldname=None,
-                 facet_class=None, boost=1.0, weight=None):
+                 facet_class=None, boost=1.0, weight=None, solr_explicit_field_type=''):
```
Comment (acdha) [inline haystack/fields.py]: Can we use None for this to match the other optional parameters?

**Comment by ahoho** [comment]
@acdha, I've made your suggested change.

I know there is limited manpower and far more important pull requests and issues to review, so I'll gladly close this to lighten the load if you think it's an unnecessary fix. My feeling was/is that, despite talk in PR #1148 that suggests the use of the Rest API, it seems like good practice to have `search_indexes.py` match whatever is going on in the schema to the extent possible.

**Comment by ahoho** [comment]
Changed to be a more general fix -- now the user can make direct alterations to any field in the Solr schema via a `solr_field_params` argument. Now additional changes to fields can be passed as a dictionary, e.g.,

 `my_field = indexes.SearchField(solr_field_params = {'type': 'string', 'sortMissingFirst': 'true'})`

**Pull request closed without merging**